### PR TITLE
Implement SPARQL 1.1 Protocol

### LINKED ISSUE #124: Support setting default-graph-uri and named-graph-uri with query
Added with #123 we read now the default-graph and named-graph parameters, which should have an effect.

### PR DESCRIPTION
**Pull request by splattater**
As described in https://www.w3.org/TR/sparql11-protocol/#protocol and https://www.w3.org/TR/sparql11-protocol/#update-operation we add protocol support for equivalents of FROM, FROM NAMED, USING, and USING NAMED.
Fixes #124

TODO

- [x] Implement SPARQL 1.1 Protocol specification
- [x] Add query rewriting
- [x] Add a method that analyzes the HTTP request and can be used for both the `/sparql` and the `/provenance` path
- [x] Disable FROM NAMED
  - [x] Return 400 and Message when used in query string
  - [x] Return 400 and Message when used in request argument
- [x] Disable USING NAMED
  - [x] Return 400 and Message when used in update string
  - [x] Return 400 and Message when used in request argument
- Add/Update tests for 
  - [x] endpoint (not needed anymore since endpoint only contains flask paths, which are tested in test_app.py
  - [x] app
  - [x] helpers
- [x] perform manual tests
- [x] Update commit messages when request arguments are used. Values of using-graph-uriwill be added, values of using-named-graph-uri woult be added but these requests will be stopped.
- [x] Check if aborted update sequences (USING NAMED) will lead to a commit if changes took place before the USING NAMED query was evaluated.
- [x] Manually review Code (squashed one commit)

### CODE REVIEW

File: quit/helpers.py [hunk 1]
```diff
@@ -119,3 +121,126 @@ def isAbsoluteUri(uri):
         return True
     else:
         return False
+
+
+def rewrite_graphs(parsed_query, default_graphs, named_graphs, sparql):
```
Comment (white-gecko) [inline quit/helpers.py]: There is not much in common for `rewrite_graphs` for update queries and for select queries. please slit these methods into methods for update and for select

Comment (splattater) [reply quit/helpers.py]: Done

**Comment by splattater** [comment]
Seperate method for query and update added.

File: quit/application.py [hunk 2]
```diff
@@ -56,8 +56,8 @@ def initialize(args):
 
     if args.garbagecollection:
         logger.info(
-            "Please use the option \"--feature garbagecollection\" instead of "
-            + "\"-gc\" or \"--garbagecollection\"."
+            "Please use the option \"--feature garbagecollection\" instead of " +
+            "\"-gc\" or \"--garbagecollection\"."
```
Comment (white-gecko) [inline quit/application.py]: Could you please undo this?

File: quit/helpers.py [hunk 3]
```diff
@@ -119,3 +124,188 @@ def isAbsoluteUri(uri):
...
+    return parsed_update
+
+
+def parse_query_type(query, base=None, default_graph=[], named_graph=[]):
+    """Parse a query and add default and named graph uri if possible."""
+    try:
+        parsed_query = parseQuery(query)
+        parsed_query = configure_query_dataset(parsed_query, default_graph, named_graph)
+        translated_query = translateQuery(parsed_query, base=base)
+        # Check if BASE is absolute http(s) URI
```
[the remark this answers is not in the document]
Comment (white-gecko) [reply quit/helpers.py]: ?

File: quit/helpers.py [hunk 4]
```diff
@@ -119,3 +124,188 @@ def isAbsoluteUri(uri):
...
+    return translated_query.algebra.name, translated_query
+
+
+def parse_update_type(query, base=None, default_graph=[], named_graph=[]):
+    """Parse an update and add default and named graph uri if possible."""
+    try:
+        parsed_update = parseUpdate(query)
+        parsed_update = configure_update_dataset(parsed_update, default_graph, named_graph)
+        translated_update = translateUpdate(parsed_update, base=base)
+        # Check if BASE is absolute http(s) URI
```
[the remark this answers is not in the document]
Comment (white-gecko) [reply quit/helpers.py]: ?

File: quit/tools/evaluate.py [hunk 5]
```diff
@@ -480,8 +480,10 @@ def evalQuery(graph, query, initBindings, base=None):
 
                 ctx.load(d.default, default=True)
 
+            # TODO re-enable original behaviour if fixed
```
Comment (white-gecko) [inline quit/tools/evaluate.py]: How will this be fixed? Please add a n issue for this if this is really necessary else remove the out commented lines.

File: quit/tools/update.py [hunk 6]
```diff
@@ -197,9 +199,13 @@ def evalModify(ctx, u):
 
                 ctx.load(d.default, default=True)
 
+            # TODO re-enable original behaviour if fixed
```
Comment (white-gecko) [inline quit/tools/update.py]: The same as with the other todo

File: quit/core.py [hunk 7]
```diff
@@ -395,14 +395,38 @@ def getFileReferenceAndContext(self, blob, commit):
...
         def build_message(message, kwargs):
             out = list()
-            for k, v in kwargs.items():
-                out.append('{}: "{}"'.format(k, v.replace('"', "\\\"")))
             if message:
-                out.append('')
                 out.append(message)
+                out.append('')
+            if query:
+                out.append('query: {}'.format(query))
```
Comment (white-gecko) [inline quit/core.py]: We expect the query to be enclosed in quotation marks `"` and also the quotation marks within the query should be escaped `\"`.

You also need to adapt the tests accordingly and you need to add a test containing a literal.

**Pull request merged**